File: 54321/reference/54321-1.0.2001.11.16/code/cube.cpp
    #include <assert.h>
    #include "cube.h"
        const unsigned int NKlein_54321::Cube::arrayLengths[] = {
            1,
            SIDE_LENGTH, 
            SIDE_LENGTH * SIDE_LENGTH, 
            SIDE_LENGTH * SIDE_LENGTH * SIDE_LENGTH,
            SIDE_LENGTH * SIDE_LENGTH * SIDE_LENGTH * SIDE_LENGTH
        };
    NKlein_54321::Cube::Cube( void )
    {
        assert( DIMENSIONS == 4 );
        CellType zero = (CellType)0;
        *this = zero;
    }
        void
        NKlein_54321::Cube::operator = (
                const NKlein_54321::Cube::CellType& value
            )
        {
                assert( &value != 0 );
            for ( unsigned int ii=0; ii < ARRAY_LEN; ++ii ) {
                this->array[ ii ] = value;
            }
        }
        void
        NKlein_54321::Cube::vectorToIndex(
                const unsigned int vec[ NKlein_54321::Cube::DIMENSIONS ],
                unsigned int* index
            )
        {
                assert( index != 0 );
                assert( vec[ 0 ] < SIDE_LENGTH );
                assert( vec[ 1 ] < SIDE_LENGTH );
                assert( vec[ 2 ] < SIDE_LENGTH );
                assert( vec[ 3 ] < SIDE_LENGTH );
                *index = 0;

                for ( unsigned int ii=0; ii < DIMENSIONS; ++ii ) {
                    *index *= SIDE_LENGTH;
                    *index += vec[ ( DIMENSIONS - 1 ) - ii ];
                }
        }
        void
        NKlein_54321::Cube::indexToVector(
                unsigned int index,
                unsigned int vec[ NKlein_54321::Cube::DIMENSIONS ]
            )
        {
                assert( index < ARRAY_LEN );
                assert( vec != 0 );
                for ( unsigned int ii=0; ii < DIMENSIONS; ++ii ) {
                    vec[ ii ] = index % SIDE_LENGTH;
                    index /= SIDE_LENGTH;
                }
        }
        bool
        NKlein_54321::Cube::determineAxis(
                unsigned int vf[ DIMENSIONS ],
                unsigned int vt[ DIMENSIONS ],
                bool wrapping,
                unsigned int* axis,
                bool* positive
            )
        {
            assert( axis != 0 );
            assert( positive != 0 );

            *positive = true;

            unsigned int diffCount = 0;
                for ( unsigned int ii=0; ii < DIMENSIONS; ++ii ) {
                    int diff = ( (int)vt[ ii ] - (int)vf[ ii ] );

                    if ( diff != 0 ) {
                        if ( wrapping ) {
                            unsigned int udiff = ( SIDE_LENGTH + diff ) % SIDE_LENGTH;
                            if ( udiff >= SIDE_LENGTH/2 ) {
                                *positive = false;
                            }
                        } else {
                            if ( diff < 0 ) {
                                *positive = false;
                            }
                        }
                        *axis = ii;
                        ++diffCount;
                    }
                }
            return ( diffCount == 1 );
        }
        NKlein_54321::Cube::CellType&
        NKlein_54321::Cube::operator [] (
                const unsigned int vec[ NKlein_54321::Cube::DIMENSIONS ]
            )
        {
            unsigned int index;
            this->vectorToIndex( vec, &index );

            return this->array[ index ];
        }
        NKlein_54321::Cube::CellType&
        NKlein_54321::Cube::operator [] (
                const unsigned int index
            )
        {
            assert( index < ARRAY_LEN );
            return this->array[ index ];
        }
        unsigned int
        NKlein_54321::Cube::getNeighbors(
                unsigned int nn[
                        2 * NKlein_54321::Cube::DIMENSIONS
                    ],
                unsigned int index,
                unsigned int dims,
                bool wrap
            )
        {
            unsigned int vec[ DIMENSIONS ];
            NKlein_54321::Cube::indexToVector( index, vec );

            unsigned int ii=0;

            if ( wrap ) {
                for ( unsigned int jj=0; jj < dims; ++jj ) {
                        unsigned int coord = vec[ jj ];
                        vec[ jj ] = ( coord + 1 ) % SIDE_LENGTH;
                        NKlein_54321::Cube::vectorToIndex( vec, &nn[ ii++ ] );
                        vec[ jj ] = ( coord + SIDE_LENGTH - 1 ) % SIDE_LENGTH;
                        NKlein_54321::Cube::vectorToIndex( vec, &nn[ ii++ ] );
                        vec[ jj ] = coord;
                }
            } else {
                for ( unsigned int jj=0; jj < dims; ++jj ) {
                        unsigned int coord = vec[ jj ];
                        if ( coord + 1 < SIDE_LENGTH ) {
                            vec[ jj ] = ( coord + 1 );
                            NKlein_54321::Cube::vectorToIndex( vec, &nn[ ii++ ] );
                        }
                        if ( coord >= 1 ) {
                            vec[ jj ] = ( coord - 1 );
                            NKlein_54321::Cube::vectorToIndex( vec, &nn[ ii++ ] );
                        }
                        vec[ jj ] = coord;
                }
            }

            return ii;
        }

File: 54321/public/src/core/cube.js
export class Cube {
  static SIDE_LENGTH = 4;
  static DIMENSIONS = 4;
  static ARRAY_LEN = 4 ** 4;
  static ARRAY_LENGTHS = [1, 4, 16, 64, 256];

  constructor() {
    this.cells = new Uint32Array(Cube.ARRAY_LEN);
  }

  fill(value = 0) {
    this.cells.fill(value);
  }

  get(index) {
    if (!Number.isInteger(index) || index < 0 || index >= Cube.ARRAY_LEN) {
      throw new RangeError(`Invalid cube index: ${index}`);
    }
    return this.cells[index];
  }

  set(index, value) {
    if (!Number.isInteger(index) || index < 0 || index >= Cube.ARRAY_LEN) {
      throw new RangeError(`Invalid cube index: ${index}`);
    }
    this.cells[index] = value;
  }

  toggle(index) {
    this.set(index, this.get(index) ^ 1);
    return this.get(index);
  }

  static indexToVector(index) {
    if (!Number.isInteger(index) || index < 0 || index >= Cube.ARRAY_LEN) {
      throw new RangeError(`Invalid cube index: ${index}`);
    }
    const vec = new Array(Cube.DIMENSIONS).fill(0);
    for (let i = 0; i < Cube.DIMENSIONS; i += 1) {
      vec[i] = index % Cube.SIDE_LENGTH;
      index = Math.floor(index / Cube.SIDE_LENGTH);
    }
    return vec;
  }

  static vectorToIndex(vec) {
    if (!Array.isArray(vec) || vec.length !== Cube.DIMENSIONS) {
      throw new TypeError('Cube vectors must contain exactly four coordinates.');
    }
    for (const coord of vec) {
      if (!Number.isInteger(coord) || coord < 0 || coord >= Cube.SIDE_LENGTH) {
        throw new RangeError(`Invalid cube coordinate: ${coord}`);
      }
    }
    let index = 0;
    for (let i = 0; i < Cube.DIMENSIONS; i += 1) {
      index *= Cube.SIDE_LENGTH;
      index += vec[Cube.DIMENSIONS - 1 - i];
    }
    return index;
  }

  static describeNeighbor(fromIndex, toIndex, dimensions, wrap = true) {
    if (!Number.isInteger(dimensions) || dimensions < 1 || dimensions > Cube.DIMENSIONS) {
      throw new RangeError(`Invalid dimension count: ${dimensions}`);
    }
    const from = Cube.indexToVector(fromIndex);

    for (let axis = 0; axis < dimensions; axis += 1) {
      const coord = from[axis];
      if (wrap || coord + 1 < Cube.SIDE_LENGTH) {
        const vec = [...from];
        vec[axis] = (coord + 1) % Cube.SIDE_LENGTH;
        if (Cube.vectorToIndex(vec) === toIndex) {
          return { axis, positive: true, dimensional: axis >= 2, wrapped: wrap && coord === Cube.SIDE_LENGTH - 1 };
        }
      }
      if (wrap || coord > 0) {
        const vec = [...from];
        vec[axis] = (coord + Cube.SIDE_LENGTH - 1) % Cube.SIDE_LENGTH;
        if (Cube.vectorToIndex(vec) === toIndex) {
          return { axis, positive: false, dimensional: axis >= 2, wrapped: wrap && coord === 0 };
        }
      }
    }
    return null;
  }

  static getNeighbors(index, dimensions, wrap = true) {
    if (!Number.isInteger(dimensions) || dimensions < 1 || dimensions > Cube.DIMENSIONS) {
      throw new RangeError(`Invalid dimension count: ${dimensions}`);
    }
    const vec = Cube.indexToVector(index);
    const neighbors = [];

    for (let axis = 0; axis < dimensions; axis += 1) {
      const coord = vec[axis];
      if (wrap) {
        vec[axis] = (coord + 1) % Cube.SIDE_LENGTH;
        neighbors.push(Cube.vectorToIndex(vec));
        vec[axis] = (coord + Cube.SIDE_LENGTH - 1) % Cube.SIDE_LENGTH;
        neighbors.push(Cube.vectorToIndex(vec));
      } else {
        if (coord + 1 < Cube.SIDE_LENGTH) {
          vec[axis] = coord + 1;
          neighbors.push(Cube.vectorToIndex(vec));
        }
        if (coord >= 1) {
          vec[axis] = coord - 1;
          neighbors.push(Cube.vectorToIndex(vec));
        }
      }
      vec[axis] = coord;
    }
    return neighbors;
  }
}
